src/tools/rbd_nbd/win32/rbd-nbd.cc

```diff
@@ -0,0 +1,1038 @@
+      f->dump_string("namespace", cfg.nsname);
+      f->dump_string("image", cfg.imgname);
+      f->dump_string("snap", cfg.snapname);
+      f->dump_int("disk_number", cfg.disk_number);
+      f->dump_string("status", status);
+      f->close_section();
+  } else {
+    should_print = true;
+    if (cfg.snapname.empty()) {
+        cfg.snapname = "-";
+    }
+    tbl << (cfg.pid || -1) << cfg.poolname << cfg.nsname
+        << cfg.imgname << cfg.snapname << cfg.devpath
+        << cfg.disk_number << status << TextTable::endrow;
+    }
+  }
+  int error = wnbd_disk_iterator.get_error();
+  if(error) {
+    derr << "Could not get disk list: " << error << dendl;
+    return -error;
+  }
+
+  if (f) {
+    f->close_section(); // devices
+    f->flush(std::cout);
+  }
+  if (should_print) {
+    std::cout << tbl;
+  }
+  if (!search_devpath.empty() && !found) {
+    return -ENOENT;
+  }
+
+  return 0;
+}
+
+static int parse_args(std::vector<const char*>& args,
+                      std::ostream *err_msg,
+                      Command *command, Config *cfg) {
+  std::string conf_file_list;
+  std::string cluster;
+  CephInitParameters iparams = ceph_argparse_early_args(
+          args, CEPH_ENTITY_TYPE_CLIENT, &cluster, &conf_file_list);
+
+  ConfigProxy config{false};
+  config->name = iparams.name;
+  config->cluster = cluster;
+
+  if (!conf_file_list.empty()) {
+    config.parse_config_files(conf_file_list.c_str(), nullptr, 0);
+  } else {
+    config.parse_config_files(nullptr, nullptr, 0);
+  }
+  config.parse_env(CEPH_ENTITY_TYPE_CLIENT);
+  config.parse_argv(args);
+  cfg->poolname = config.get_val<std::string>("rbd_default_pool");
+
+  std::vector<const char*>::iterator i;
+  std::ostringstream err;
+
+  for (i = args.begin(); i != args.end(); ) {
+    if (ceph_argparse_flag(args, i, "-h", "--help", (char*)NULL)) {
+      return HELP_INFO;
+    } else if (ceph_argparse_flag(args, i, "-v", "--version", (char*)NULL)) {
+      return VERSION_INFO;
+    } else if (ceph_argparse_witharg(args, i, &cfg->devpath, "--device", (char *)NULL)) {
+    } else if (ceph_argparse_witharg(args, i, &cfg->format, err, "--format",
+                                     (char *)NULL)) {
+    } else if (ceph_argparse_flag(args, i, "--pretty-format", (char *)NULL)) {
+      cfg->pretty_format = true;
+    } else if (ceph_argparse_witharg(args, i, &cfg->parent_pipe, err, "--pipe-handle", (char *)NULL)) {
+      if (!err.str().empty()) {
+        *err_msg << "rbd-nbd: " << err.str();
+        return -EINVAL;
+      }
+      if (cfg->parent_pipe < 0) {
+        *err_msg << "rbd-nbd: Invalid argument for pipe-handle!";
+        return -EINVAL;
+      }
+    } else {
+      ++i;
+    }
+  }
+
+  Command cmd = None;
+  if (args.begin() != args.end()) {
+    if (strcmp(*args.begin(), "map") == 0) {
+      cmd = Connect;
+    } else if (strcmp(*args.begin(), "unmap") == 0) {
+      cmd = Disconnect;
+    } else if (strcmp(*args.begin(), "list-mapped") == 0) {
```

Comment:
Have those commands been deprecated? If not, we'd like stay as close as possible to the original CLI so that any 3rd party automation built around rbd-nbd can easily support Windows as well. It's also easier for the users that might be used to those commands or users that are following the docs.

Talking about automation, I sent a couple of patches on OpenStack so that Hyper-V may use Ceph volumes:
https://review.opendev.org/#/c/718399/
https://review.opendev.org/#/c/718403/

We might consider Kubernetes as well.

Reply:
Our documentation says to use via the rbd CLI -- but people skip that step. This is a new utility so it doesn't need to be backwards compatible w/ the Linux versions. If there is a way to share testing between Linux + Windows (unlikely in the nearterm), the tests would just use "rbd device map/unmap/...".

Reply:
FWIW, we did run some of the existing tests: https://github.com/ceph/ceph/files/4077517/test_results.zip

Those are mostly unit tests, so IIRC those weren't actually testing the commands. Still providing useful coverage nevertheless, helped us find a few bugs.

Reply:
In a perfect world, we will get a Jenkins builder for the Win32 environment so that with each `ceph-ci` build we can immediately tell if some change has broken the win32 build. Without such a plan, I suspect it won't take long for a random commit to break the build. 

Later we can create a Windows VM w/ a sysprep or similar script to install WNBD + rbd/rbd-nbd and run some basic integration tests as part of the RBD suite (presuming Microsoft licensing allows for that).

Reply:
Yes, making sure that it still builds would be a great start. At the moment, we're building it using Mingw on Ubuntu or OpenSuse, so this part doesn't require a Windows environment.

Windows trial images can be used when actually running tests. We've been doing it with the OpenStack Hyper-V CI for quite some time now.

Reply:
Awesome. If you have a script to use Mingw on Ubuntu, perhaps you can open a PR against ceph-build to include that script [1] for all ceph-ci based builds. 

[1] https://github.com/ceph/ceph-build/tree/master/ceph-dev-new-build/build

Reply:
> Awesome. If you have a script to use Mingw on Ubuntu, perhaps you can open a PR against ceph-build to include that script [1] for all ceph-ci based builds.
> 
> [1] https://github.com/ceph/ceph-build/tree/master/ceph-dev-new-build/build

That's great, thanks!

> Nit: let's drop list-mapped (and now show-mapped) below since this is a new utility and we don't need to worry about backwards compatibility.
> Our documentation says to use via the rbd CLI -- but people skip that step. This is a new 
utility so it doesn't need to be backwards compatible w/ the Linux versions. If there is a way to share testing between Linux + Windows (unlikely in the nearterm), the tests would just use "rbd device map/unmap/...".

Double checked, ``list-mapped`` is actually used by "rbd device list", so I think you actually meant dropping the ``list`` command, which we added for convenience.

Reply:
Since we're using a separate wrapper now (Wnbd.cc), I've updated it to just use "list" instead of "list-mapped" as suggested.